daemon/fqdn.go

```diff
@@ -357,7 +357,7 @@ func (d *Daemon) bootstrapFQDN(restoredEndpoints *endpointRestoreState, preCache
 			)
 			record.Log()
 
-			if msg.Response && msg.Rcode == dns.RcodeSuccess {
+			if msg.Response && msg.Rcode == dns.RcodeSuccess && len(responseIPs) > 0 {
```

Comment:
What exactly are the semantics behind no IPs being returned here? Could that mean that there are no longer any IPs to which this DNS name resolves, and thus we should regenerate?

Reply:
@ianvernon 

>There are cases where a DNS response return a Nonerror but there are no
IPS in the response. In that case Cilium regenerates the policy because
it's a legit response.
>
>With this change, regeneration will not happens if it is a empty
response.

Reply:
@aanm I guess I was asking how a non-error but no IPs being resolved could happen. Wouldn't no IPs being resolved indicate that the host isn't reachable? It just seems like a weird case to me.

Reply:
If no IPS are in the dns response we don't trigger the policyAdd due there is no new identities, so there is no reason to call PolicyAdd again.

Reply:
@eloycoto hold on a second but if there are no IPs, should we allow or deny everything?

Reply:
Historically, we relied on each new DNS lookup to update what was in the datapath. We've since introduced a cache (https://github.com/cilium/cilium/blob/master/pkg/fqdn/cache.go) and a TTL expirer (https://github.com/cilium/cilium/blob/master/daemon/fqdn.go#L140-L175). Between the two of these seeing no IPs doesn't mean much, since any expired entries will already have been removed from the datapath.
More generally, it will be common to see no IPs for AAAA lookups so we can't treat a single response as "block all" or "allow all". The collected IPs we've seen from any DNS response for a DNS name are the allowed IPs for it (accounting for TTL expiration).

Reply:
Ok, but what happens if we add a toFQDN policy and an empty response is received from the DNS request? Doesn't this mean we will allow everything because we are now skipping policy regeneration of an empty response?

Reply:
There was a period of time where we did that accidentally but it was a bug. When a ToFQDNs rule is present we enforce policy and if no IPs are seen we would deny everything. We would also deny traffic until the first IPs are seen (from one of the DNS proxy, the DNS poller, or the pre-cache file that preseeds the DNS data). The policy enforcement part isn't directly handled by this code but by the normal policy stuff by calling https://github.com/cilium/cilium/blob/master/pkg/policy/api/egress.go#L160-L168 and https://github.com/cilium/cilium/blob/master/pkg/policy/api/fqdn.go#L94-L96.
Basically, a ToFQDNs rule is enforces L3 policy and denies all, but then the toCIDR rules generated whitelist specific IPs. This PR changes when we would update those IPs in a roundabout way. It won't update the rulegen/cache unless there are IPs in the response. This means rulegen won't recreate any rules. Previously, it would have updated rulegen and that might have recreated rules if something expired in the meantime. This shouldn't happen because the TTL expirer would have already done this (but there is a race).

None of this is very self-contained :(